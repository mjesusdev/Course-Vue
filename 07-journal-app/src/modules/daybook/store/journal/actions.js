/* export const myAction = async ({ commit }) => {

} */

export const loadEntries = async() => {
    
}

export const updateEntry = async() => {

}

export const createEntry = async() => {

}