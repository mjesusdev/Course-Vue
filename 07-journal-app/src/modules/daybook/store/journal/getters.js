/* export const myGetter = ( state ) => {
    return state
} */

export const getEntriesByTerm = ( state ) => ( term = '' ) => {
    if ( term.length === 0 )  return state.entries

    return state.entries.filter( entry => entry.text.toLowerCase().includes( term.toLocaleLowerCase() ) )
}

export const getEntryById = () => {
    
}