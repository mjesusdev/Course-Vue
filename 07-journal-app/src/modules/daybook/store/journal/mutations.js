/* export const myAction = ( state ) => {

} */

export const setEntries = () => {

}

export const updateEntry = () => {

}

export const addEntry = () => {

}