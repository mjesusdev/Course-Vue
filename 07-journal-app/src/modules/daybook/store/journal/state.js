export default () => ({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(), // 123456789
            date: new Date().toDateString(), // thursday 21 October
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro ea accusantium, sequi saepe quis obcaecati pariatur aut, molestiae dolor in rerum sit distinctio maiores omnis asperiores tempora laudantium! Nemo, ullam.',
            picture: null // https
        },
        {
            id: new Date().getTime() + 1000, // 1234567893232
            date: new Date().toDateString(), // thursday 21 October
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro ea accusantium, sequi saepe quis obcaecati pariatur aut, molestiae dolor in rerum sit distinctio maiores omnis asperiores tempora laudantium! Nemo, ullam.',
            picture: null // https
        },
        {
            id: new Date().getTime() + 2000, // 1234567893232234
            date: new Date().toDateString(), // thursday 21 October
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro ea accusantium, sequi saepe quis obcaecati pariatur aut, molestiae dolor in rerum sit distinctio maiores omnis asperiores tempora laudantium! Nemo, ullam.',
            picture: null // https
        }
    ]
})