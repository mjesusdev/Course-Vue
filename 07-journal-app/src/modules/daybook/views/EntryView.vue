<template>
    <div>
        <template v-if="entry">
            <div class="entry-title d-flex justify-content-between p-2">
                <div>
                    <span class="text-success fs-3 fw-bold">{{ day }}</span>
                    <span class="mx-1 fs-3">{{ month }}</span>
                    <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
                </div>

                <div>
                    <button class="btn btn-danger mx-2">
                        Delete
                        <i class="fa fa-trash-alt"></i>
                    </button>

                    <button class="btn btn-primary">
                        Upload Photo
                        <i class="fa fa-upload"></i>
                    </button>
                </div>
            </div>

            <hr />

            <div class="d-flex flex-column px-3 h-75">
                <textarea 
                    rows="30"
                    v-model="entry.text" 
                    placeholder="¿Qué sucedió hoy?"
                ></textarea>
            </div>

            <img src="https://bit.ly/3naP2Z2" class="p-3" alt="Img StarWars">
        </template>

        <Fab
            icon="fa-save"
            @on:click="saveEntry"
        />

    </div>
</template>

<script>
    import { defineAsyncComponent } from 'vue'
    import { mapGetters, mapActions } from 'vuex'

    import getDayMonthYear from '../helpers/getDayMonthYear';

    export default {
        props: {
            id: {
                type: String,
                required: true
            }
        },

        components: {
            Fab: defineAsyncComponent( () => import('../components/Fab.vue') )
        },

        data() {
            return {
                entry: {
                    text: ''
                }
            }
        },

        computed: {
            ...mapGetters('journal', ['getEntryById']),
            day() {
                const { day } = getDayMonthYear( this.entry.date )
                return day
            },
            month() {
                const { month } = getDayMonthYear( this.entry.date )
                return month
            },
            yearDay() {
                const { yearDay } = getDayMonthYear( this.entry.date )
                return yearDay
            }
        },

        methods: {
            ...mapActions('journal', ['updateEntry']),

            loadEntry() {
                const entry = this.getEntryById( this.id )
                if ( !entry ) return this.$router.push({ name: 'no-entry' })

                this.entry = entry
            },

            async saveEntry() {
                console.log('Guardando entrada');

                // Call Action - Journal Module
                this.updateEntry( this.entry )
            }
        },

        created() {
            this.loadEntry()
        },

        watch: {
            id() {
                this.loadEntry()
            }
        }
    }
</script>

<style lang="scss" scoped>
    textarea {
        font-size: 20px;
        border: none;
        height: 100%;

        &:focus {
            outline: none;
        }
    }

    img {
        width: 200px;
        position: fixed;
        bottom: 150px;
        right: 20px;
        box-shadow: 0px 5px 10px rgba($color: #000, $alpha: 0.2);
    }
</style>