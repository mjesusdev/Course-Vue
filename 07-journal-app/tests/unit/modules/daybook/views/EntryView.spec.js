import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import Swal from 'sweetalert2'

import journal from '@/modules/daybook/store/journal'
import { journalState } from '../../../mock-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView'

const createVuexStore = ( initialState ) => 
    createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...initialState }
            }
        }
    })

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Tests in EntryView', () => {
    
    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }
 
    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView , {
            props: {
                id: '-Mn5HXnHBBwU8rtrMPZ8'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })

    test('should exit user when ID not exists', () => {
        const wrapper = shallowMount( EntryView , {
            props: {
                id: 'ID not exists in store'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
        
        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })
    })

    test('should show entries', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()
    })

    test('should delete entries and exit', (done) => {
        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }) )
        wrapper.find('.btn-danger').trigger('click')

        expect( Swal.fire ).toHaveBeenCalledWith({     
            title: 'Â¿Are you sure?',
            text: 'If you delete this, not recovery',
            showDenyButton: true,
            confirmButtonText: "Yes, I'm sure"
        })

        setTimeout( () => {
            expect( store.dispatch ).toHaveBeenCalledWith('journal/deleteEntry', '-Mn5HXnHBBwU8rtrMPZ8')
            expect( mockRouter.push ).toHaveBeenCalled()
            done()
        }, 1)

    })

})