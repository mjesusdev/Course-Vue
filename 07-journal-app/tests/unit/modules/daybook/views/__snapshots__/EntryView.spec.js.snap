// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in EntryView should show entries 1`] = `
<div>
  <div class="entry-title d-flex justify-content-between p-2">
    <div><span class="text-success fs-3 fw-bold">28</span><span class="mx-1 fs-3">Octubre</span><span class="mx-2 fs-4 fw-light">2021, Jueves</span></div>
    <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload Photo <i class="fa fa-upload"></i></button></div>
  </div>
  <hr>
  <div class="d-flex flex-column px-3 h-75"><textarea rows="30" placeholder="¿Qué sucedió hoy?"></textarea></div>
  <!--v-if-->
  <!--v-if-->
  <fab-stub icon="fa-save"></fab-stub>
</div>
`;
